File: upl-purchase-sdk-php/src/Api/BaseApi.php
<?php

/**
 * Purchase API
 * @copyright Copyright (c) 2020 Unzer Pay Later
 * @license see LICENSE.TXT
 */

declare(strict_types=1);

namespace Unzer\PayLater\Api;

use Unzer\PayLater\Communication\Communicator;
use Unzer\PayLater\Communication\ResponseBuilder;
use Unzer\PayLater\Communication\ResponseHeaderCollection;
use Unzer\PayLater\Exception\ApiResponseException;
use Unzer\PayLater\Exception\AuthorizationException;
use Unzer\PayLater\Exception\BuilderException;
use Unzer\PayLater\Exception\ReferenceException;
use Unzer\PayLater\Exception\ResponseException;
use Unzer\PayLater\Exception\ServerErrorException;
use Unzer\PayLater\Exception\ValidationException;
use Unzer\PayLater\Model\OperationResult;
use Unzer\PayLater\Model\PurchaseOperationResponse;
use Throwable;

use function json_decode;
use function json_last_error;
use function sprintf;
use function str_replace;

use const JSON_ERROR_NONE;

abstract class BaseApi
{
    /**
     * @var Communicator
     */
    protected $communicator;

    /**
     * @param Communicator $communicator
     */
    public function __construct(Communicator $communicator)
    {
        $this->communicator = $communicator;
    }

    /**
     * @param string $uri
     * @param string $key
     * @param string $value
     * @return string
     */
    protected function populateUri(string $uri, string $key, string $value): string
    {
        $key = sprintf('{%s}', $key);
        return str_replace($key, $value, $uri);
    }

    /**
     * @param ResponseException $responseException
     * @return ApiResponseException
     * @throws BuilderException
     */
    protected function createApiResponseException(ResponseException $responseException): ApiResponseException
    {
        $purchaseOperationResponse = null;
        if ($this->isJsonResponse($responseException->getBody()) === true) {
            $purchaseOperationResponse = ResponseBuilder::build(
                PurchaseOperationResponse::class,
                json_decode($responseException->getBody(), true)
            );
        }
        $operationResult = $purchaseOperationResponse instanceof PurchaseOperationResponse ?
            $purchaseOperationResponse->getResult() :
            null;

        return $this->getApiResponseException(
            $responseException->getStatusCode(),
            $responseException->getBody(),
            $responseException->getResponseHeaderCollection(),
            $operationResult,
            $responseException->getPrevious()
        );
    }

    /**
     * @param string $responseBody
     * @return bool
     */
    private function isJsonResponse(string $responseBody): bool
    {
        json_decode($responseBody, true);
        return (json_last_error() === JSON_ERROR_NONE);
    }

    // phpcs:disable Generic.Metrics.CyclomaticComplexity.TooHigh
    // phpcs:disable ObjectCalisthenics.Files.FunctionLength.ObjectCalisthenics\Sniffs\Files\FunctionLengthSniff

    /**
     * @param int $statusCode
     * @param string $responseBody
     * @param ResponseHeaderCollection $responseHeaders
     * @param OperationResult|null $operationResult,
     * @param Throwable|null $previous
     * @return ApiResponseException
     */
    private function getApiResponseException(
        int $statusCode,
        string $responseBody,
        ResponseHeaderCollection $responseHeaders,
        ?OperationResult $operationResult,
        ?Throwable $previous = null
    ): ApiResponseException {
        $errorId = $operationResult instanceof OperationResult ? $operationResult->getStatusCode() : null;
        $message = $operationResult instanceof OperationResult ? $operationResult->getStatusMessage() : null;
        switch ($statusCode) {
            case 400:
                return new ValidationException(
                    $statusCode,
                    $responseBody,
                    $responseHeaders,
                    $errorId,
                    $message,
                    $operationResult,
                    $previous
                );
            case 401:
            case 403:
                return new AuthorizationException(
                    $statusCode,
                    $responseBody,
                    $responseHeaders,
                    $errorId,
                    $message,
                    $operationResult,
                    $previous
                );
            case 404:
            case 409:
            case 410:
                return new ReferenceException(
                    $statusCode,
                    $responseBody,
                    $responseHeaders,
                    $errorId,
                    $message,
                    $operationResult,
                    $previous
                );
            case 500:
            case 502:
            case 503:
                return new ServerErrorException(
                    $statusCode,
                    $responseBody,
                    $responseHeaders,
                    $errorId,
                    $message,
                    $operationResult,
                    $previous
                );
            default:
                $message = $message !== null ? $message : '';
                return new ApiResponseException(
                    $message,
                    $statusCode,
                    $responseBody,
                    $responseHeaders,
                    $message,
                    $operationResult,
                    $previous
                );
        }
    }

    // phpcs:enable ObjectCalisthenics.Files.FunctionLength.ObjectCalisthenics\Sniffs\Files\FunctionLengthSniff
    // phpcs:enable Generic.Metrics.CyclomaticComplexity.TooHigh
}

File: upl-purchase-sdk-java/src/main/java/com/unzer/paylater/api/BaseApi.java
package com.unzer.paylater.api;

import java.io.Closeable;
import java.io.IOException;
import java.util.List;

import com.unzer.paylater.communication.Communicator;
import com.unzer.paylater.communication.RequestHeader;
import com.unzer.paylater.exception.ApiException;
import com.unzer.paylater.exception.AuthorizationException;
import com.unzer.paylater.exception.UnzerException;
import com.unzer.paylater.exception.ReferenceException;
import com.unzer.paylater.exception.ResponseException;
import com.unzer.paylater.exception.ValidationException;
import com.unzer.paylater.logging.CommunicatorLogger;
import com.unzer.paylater.model.OperationResult;
import com.unzer.paylater.model.PurchaseOperationResponse;

/**
 * Base class of all Unzer Pay Later API resources.
 */
public abstract class BaseApi implements Closeable {

    protected final Communicator communicator;

    protected BaseApi(Communicator communicator) {
        if (communicator == null) {
            throw new IllegalArgumentException("communicator is required");
        }
        this.communicator = communicator;
    }

    public Communicator getCommunicator() {
        return communicator;
    }

    public void enableLogging(CommunicatorLogger logger) {
        communicator.enableLogging(logger);
    }
    public void disableLogging() {
        communicator.disableLogging();
    }

    protected String populateUri(String uri, String key, String value) {
        return uri.replace("{" + key + "}", value);
    }

    protected void addHeaderParam(List<RequestHeader> headerParams, String headerName, String headerValue) {
        if (headerValue != null && !headerValue.trim().isEmpty()) {
            headerParams.add(new RequestHeader(headerName, headerValue));
        }
    }

    protected RuntimeException createException(ResponseException e) {
        String responseBody = e.getBody();
        PurchaseOperationResponse errorObject = communicator.unmarshal(responseBody, PurchaseOperationResponse.class);
        OperationResult result = errorObject != null ? errorObject.getResult() : null;

        switch (e.getResponseStatusCode()) {
            case 400:
                return new ValidationException(e, result);
            case 401:
            case 403:
                return new AuthorizationException(e, result);
            case 404:
            case 409:
            case 410:
                return new ReferenceException(e, result);
            case 500:
            case 502:
            case 503:
                return new UnzerException(e, result);
            default:
                return new ApiException(e, result);
        }
    }

    @Override
    public void close() throws IOException {
        this.communicator.close();
    }
}
